### PR TITLE
UI improvements for holsters

### PR DESCRIPTION
**Pull request by mugling**
Items now display compatible holsters/sheathes etc:

s40

For holsters the raw stats are also displayed (and can be compared):

hols

### CODE REVIEW

**Comment by mugling** [comment]
Note a284f1b fixes a mistake I made with the (as yet incomplete) flag documentation tool to avoid a merge conflict when I add documentation for `BELT_CLIP`, `SHEATH_KNIFE` etc.

The code support is complete - see screenshots in updated preamble.

It is inconsistent which boots can store knives (for example rubber boots can but motorcycle can't). The only reason we still have boots storing knives is because we had to preserve save compatibility but I have now a solution that avoids that problem. I'd like to reduce the number for both balance reasons and because the above interface is ugly when there are two dozen holsters (this problem only occurs with boots).

Some variety would be good - I'm thinking combat boots should be `volume 1` and `knee high boots` a pair of `volume 2`. We should probably give throwing knives `SHEATH_KNIFE` to make selection of appropriate boots more relevant for players that want to use them.

Also any other suggestions what items could go on `BELT_CLIP`. Possibly heavy duty flashlight?

**Comment by mugling** [comment]
As we increase the variety for holsters:

western

The extra UI is helpful in providing suggestions:

**Comment by Rivet-the-Zombie** [comment]
Cowboy boots don't have integrated holsters. That would throw off their goofy looking aesthetic.

**Comment by mugling** [comment]
How about if the wording was changed to say there was enough room to conceal a small gun inside them:

img_1020

**Comment by Rivet-the-Zombie** [comment]
Yeah it'd have to allow storing one in each boot. This also has me thinking that we need tac holsters that can hold a gun plus a spare mag.

**Comment by mugling** [comment]
> This also has me thinking that we need tac holsters that can hold a gun plus a spare mag.

Yes - this already works in code - and a UI implementation isn't far off

**Comment by Rivet-the-Zombie** [comment]
Ausgezeichnet!

**Comment by mugling** [comment]
More suggestions for boots...

File: src/item.cpp [hunk 1]
```diff
@@ -1635,6 +1630,24 @@ std::string item::info( bool showtext, std::vector<iteminfo> &info ) const
...
+            return ptr->can_holster( *this );
+        } );
+
+        if( !holsters.empty() ) {
+            std::vector<std::string> opts;
+            std::transform( holsters.begin(), holsters.end(), std::back_inserter( opts ), []( const itype *e ) {
+                return e->nname( 1 );
+            } );
+            insert_separation_line();
+            info.emplace_back( "DESCRIPTION", _( "<bold>Can be stored in:</bold> " ) + enumerate_as_string( opts ) );
```
Comment (codemime) [inline src/item.cpp]: You don't need a temporary vector for this. There's an overloaded version of `enumerate_as_string()`:
```c++
enumerate_as_string( holsters.begin(), holsters.end(), []( const itype *e ) {
    return e->nname( 1 );
} );
```

File: src/item_factory.h [hunk 2]
```diff
@@ -238,6 +242,20 @@ class Item_factory
         const std::map<const itype_id, std::unique_ptr<itype>> &get_all_itypes() const {
             return m_templates;
         }
+
+        /** Find all templates matching the UnaryPredicate function */
+        static std::vector<const itype *> find( const std::function<bool( const itype & )> &func ) {
+            std::vector<const itype *> res;
+            if( item_controller ) {
```
Comment (codemime) [inline src/item_factory.h]: Can this condition ever be `false`?

Comment (mugling) [inline src/item_factory.h]: Only if called before `load_core_data`

**Comment by Rivet-the-Zombie** [comment]
Also, it occurs to me that putting stuff in boots should increase their encumbrance a good bit while it's in there.

File: src/item_factory.h (hunk 2 above)
Comment (codemime) [inline src/item_factory.h]: > Perhaps the check is overly paranoid?

I guess so. `item_controller` is a global variable:
```c++
std::unique_ptr<Item_factory> item_controller( new Item_factory() );
```
It should either be instantiated  before any code can call it from the main cycle, or throw an exception if `new` fails due to lack of free memory.

**Comment by flitvious** [comment]
Although I agree with @Rivet-the-Zombie about boots encumbrance, this would create difference between two kinds of equipment. First are boots, belts and holsters that can have items inside them. Second are backpacks, duffel bags and general gear with storage that add to one big virtual "volume". Right now both kinds of equipment have the same constant encumbrance when full and when empty.

**Comment by mugling** [comment]
> Also, it occurs to me that putting stuff in boots should increase their encumbrance a good bit while it's in there.

This can probably come in a later PR

**Comment by mugling** [comment]
Ready

**Comment by Coolthulhu** [comment]
Minor: units are displayed in liters, not in the configured unit.

**Comment by Coolthulhu** [comment]
OK, that's a problem:
* Spawn ankle holster
* Change weight settings to lbs
* "It can store 2.20 kg"

The technical parts seem to work fine, the unit conversion is wrong.

**Comment by mugling** [comment]
Fixed

**Comment by mugling** [comment]
Any chance of a retest on this? It's a conflict with anything else that updates `flags.json` and also anything that uses `data/legacy` migration. I have a dependent branch for the latter that overhauls vehicle movement and fuel consumption

File: src/iuse_actor.cpp [hunk 3]
```diff
@@ -2003,11 +2003,12 @@ long holster_actor::use( player *p, item *it, bool, const tripoint & ) const
 
 void holster_actor::info( const item&, std::vector<iteminfo>& dump ) const
 {
-    dump.emplace_back( "TOOL", _( "Can contain items up to " ), "<num> L",
-                       round_up( units::to_liter( max_volume ), 1 ), false, "", max_weight <= 0 );
+    int scale = 0;
+    dump.emplace_back( "TOOL", _( "Can contain items up to " ), string_format( "<num> %s", volume_units_abbr() ),
+                       round_up( convert_volume( max_volume.value(), &scale ), 1 ), scale == 0, "", max_weight <= 0 );
```
Comment (Coolthulhu) [inline src/iuse_actor.cpp]: Round up sounds misleading. It should probably round down instead.

Comment (mugling) [inline src/iuse_actor.cpp]: Yes, but we don't have a `round_down` function yet. Could you tag that as an issue though and I'll write that function in a separate PR and also check for other suspicious calls to `round_up` in the codebase at the same time?

**Comment by Coolthulhu** [comment]
Totally unrelated to holsters themselves, but the version bumping thing could use a rework.
What you want here is not a single number, but a set of flags like "MIGRATED_HOLSTERS", "MIGRATED_TOOLMODS" etc.
Otherwise one migrating PR locks out all the others.

**Comment by mugling** [comment]
Yes, that could work well actually. Once this is through I'll look into that

**Comment by Coolthulhu** [comment]
I'm currently on a different branch. Did you test it? If it fixes at least that kg->lb thing, then it should be fine to merge.

**Comment by mugling** [comment]
Yes, I tested it (for example you can have quarts and pounds)

**Comment by mugling** [comment]
So can that go in then?

**Pull request merged**